File: 421.rb
=begin



n^15 + 1 = (n+1) (n^4-n^3+n^2-n+1) (n^10-n^5+1)

n^15 + 1




=end

require './euler_lib.rb'
#Primes.setup(10**8)

def s(n, bound)
  v = n**15 + 1
  factors = Array.new
  factors.push (n+1)
  v /= (n+1)
  factors.push (n**4 - n**3 + n**2 - n + 1)
  v /= (n**4 - n**3 + n**2 - n + 1)
  factors.push(n**10 - n**5 + 1)
  v /= (n**10 - n**5 + 1)
  return factors
end

pp s(10, 10**12)

File: 421.cpp
#include <iostream>
#include <bitset>
#include <cmath>
#include <vector>

using namespace std;

#define MAX_PRIME 100000000
#define ulong unsigned long long


class Primes {
  private:
    bitset<MAX_PRIME> *b;
  public:
		vector<ulong> *primes;
    Primes();
    int is_prime(ulong);
};


Primes::Primes() {
  b =  new std::bitset<MAX_PRIME>;
	primes =  new std::vector<ulong>;
  b->reset();
  b->flip();
  b->set(0,0);
  for (int i=2; i < MAX_PRIME; i++) {
    if (b->test(i)) {
			primes->push_back(i);
      ulong j = i*i;
      while (j < MAX_PRIME) {
        b->set(j,0);
        j += i;
      }
    }
  }
	cout << "Primes are setup" << endl;
  return;  
}

int Primes::is_prime(ulong n) {
  return(b->test(n) == 1);
};


int main() {
  Primes p;
  return(0);
}
